File: apps/api/app/schemas/role.py
from pydantic import BaseModel
from typing import Optional, List
from .permission import PermissionRead

class RoleBase(BaseModel):
    name: str
    description: Optional[str] = None
    is_system: bool = False
    is_active: bool = True

class RoleCreate(RoleBase):
    permission_ids: List[int] = []

class RoleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    permission_ids: Optional[List[int]] = None

class RoleRead(RoleBase):
    id: int
    permissions: List[PermissionRead] = []
    
    class Config:
        from_attributes = True

class PermissionAuditLogRead(BaseModel):
    id: int
    actor_id: int
    role_id: int
    action: str
    permission_name: str
    timestamp: str
    actor_name: Optional[str] = None
    role_name: Optional[str] = None

    class Config:
        from_attributes = True

File: apps/web/src/types/role.ts
export interface Permission {
    id: number;
    name: string;
    description?: string;
    module: string;
}

export interface PermissionGroup {
    module: string;
    permissions: Permission[];
}

export interface Role {
    id: number;
    name: string;
    description?: string;
    is_system: boolean;
    is_active: boolean;
    permissions: Permission[];
}

export interface RoleCreateDTO {
    name: string;
    description?: string;
    permission_ids: number[];
}

export interface RoleUpdateDTO {
    name?: string;
    description?: string;
    is_active?: boolean;
    permission_ids?: number[];
}

export interface PermissionAuditLog {
    id: number;
    actor_id: number;
    actor_name?: string;
    role_id: number;
    role_name?: string;
    action: 'ADD_PERMISSION' | 'REMOVE_PERMISSION';
    permission_name: string;
    timestamp: string;
}
